lib/extractor/image.ts

```diff
@@ -0,0 +1,145 @@
+export { ImageName, ImageDigest, DigestType };
+class ImageName {
+  public name: string;
+  public tag?: string;
+  public digests: Record<string, ImageDigest>;
+
+  constructor(
+    targetImage: string,
+    digests: { manifest?: string; index?: string } = {},
+  ) {
+    // this regex has been copied from
+    // https://github.com/distribution/distribution/blob/fb2188868d771aa27e5781a32bf78d4c113c18a6/reference/regexp.go#L101
+    // (code has been modified to print the regex), and then adjusted to
+    // Javascript. The required modifications were replacing `[:xdigit:]` with
+    // `[a-fA-F0-9]` and escaping the slashes.
+    // Note that the digest matched in this Regex will match digests that have
+    // uppercase-letters, while the regex used in validateDigest does NOT match
+    // uppercase-letters. This simply matches the behaviour from the upstream
+    // `reference` and `go-digest `packages.
+    //
+    // we're matching pattern: <registry:port_number>(optional)/<image_name>(mandatory):<image_tag>(optional)@<tag_identifier>(optional)
+    // This Regex contains three capture groups:
+    // 1) The repository / image name
+    // 2) tag
+    // 3) digest
+    const re =
+      /^((?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?\/)?[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:\/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][A-Fa-f0-9]{32,}))?$/gi;
+
+    const groups = re.exec(targetImage);
+    if (groups === null) {
+      if (targetImage === "") {
+        throw Error("image name is empty");
+      }
+      if (re.exec(targetImage.toLowerCase()) !== null) {
+        throw Error("image repository contains uppercase letter");
+      }
+      throw Error("invalid image reference format");
+    }
+
+    const GROUP = {
+      NAME: 1,
+      TAG: 2,
+      DIGEST: 3,
+    };
```

Comment:
How about doing this instead:
```js
const parsed = {
  name: groups[1],
  tag: groups[2],
  digest: groups[3],
}
```

That way we don't need to constantly index the `group` and can instead directly access the elements, e.g.
```
if (parsed.name.length > NameTotalLengthMax) { ... }
this.name = parsed.name
this.tag = parsed.tag
```

We could also go further and directly assign the name & tag variable to the `this` fields:
```js
    this.name = groups[1];
    this.tag = groups[2];
    const parsedDigest = groups[3];

    const NameMaxLength = 255;
    if (this.name.length > NameMaxLength) {
      throw Error("image repository name is more than 255 characters");
    }

    if (!this.tag && !parsedDigest) {
      this.tag = "latest";
    }
```
(the only thing that we can't directly assign is the digest)